### PR TITLE
Android: Implement SQLite-based game library.

### PR DESCRIPTION
**Pull request by sigmabeta**
The user can now select some folders that contain games; those folders are scanned and metadata from the games is saved into an SQLite database (so the native methods for getting that metadata don't need to be used again.)

No method is currently in place to refresh or clear the library; I'd like to put those in, but it would almost certainly create a nasty merge conflict with #2413, so I'd rather wait until that's merged.

There are also some other, less sexy changes: I got rid of the Game->WiiGame/GcGame/etc model, because the inheritance between these is already handled pretty well on the native side; some UI optimizations; and now the cards on the grid show the game's Maker instead of a highly-likely-to-be-cut-off description.

### CODE REVIEW

**Comment by sigmabeta** [comment]
Not sure what the deal is with that Android build failure - it's referencing code that was definitely deleted. Maybe pr-android is having trouble because the last PR I made has changes to how the app is built (which are not included in this PR)?

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameAdapter.java [hunk 1]
```diff
@@ -13,25 +16,31 @@
...
 import org.dolphinemu.dolphinemu.emulation.EmulationActivity;
-import org.dolphinemu.dolphinemu.model.Game;
+import org.dolphinemu.dolphinemu.model.GameDatabase;
 import org.dolphinemu.dolphinemu.viewholders.GameViewHolder;
 
-import java.util.ArrayList;
-
-public class GameAdapter extends RecyclerView.Adapter<GameViewHolder> implements
+/**
+ * This adapter, unlike the FileAdapter which is backed by an ArrayList, gets its
```
Comment (lioncash) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameAdapter.java]:
```java
* This adapter, unlike the {@link FileAdapter}
```

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameAdapter.java [hunk 2]
```diff
@@ -158,9 +264,24 @@ public void getItemOffsets(Rect outRect, View view, RecyclerView parent, Recycle
 		}
 	}
 
-	public void setGameList(ArrayList<Game> gameList)
+	private class GameDataSetObserver extends DataSetObserver
```
Comment (lioncash) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameAdapter.java]: final class.

**Comment by sigmabeta** [comment]
@dolphin-emu-bot rebuild

**Comment by sigmabeta** [comment]
@dolphin-emu-bot rebuild

**Pull request merged**